File: algorithms/code/tests/binary_search_tree_test.go
package main

import (
	"reflect"
	"testing"
)

func TestBinarySearchTree_Insert(t *testing.T) {
	bst := NewBinarySearchTree()
	
	// Test inserting into empty tree
	bst.Insert(50)
	if bst.Root == nil || bst.Root.Value != 50 {
		t.Errorf("Expected root value 50, got %v", bst.Root)
	}
	
	// Test inserting smaller value
	bst.Insert(30)
	if bst.Root.Left == nil || bst.Root.Left.Value != 30 {
		t.Errorf("Expected left child 30, got %v", bst.Root.Left)
	}
	
	// Test inserting larger value
	bst.Insert(70)
	if bst.Root.Right == nil || bst.Root.Right.Value != 70 {
		t.Errorf("Expected right child 70, got %v", bst.Root.Right)
	}
	
	// Test inserting duplicate (should not change tree)
	originalSize := bst.Size()
	bst.Insert(50)
	if bst.Size() != originalSize {
		t.Errorf("Expected size to remain %d after duplicate insert, got %d", originalSize, bst.Size())
	}
}

func TestBinarySearchTree_Search(t *testing.T) {
	bst := NewBinarySearchTree()
	values := []int{50, 30, 70, 20, 40, 60, 80}
	
	// Insert values
	for _, value := range values {
		bst.Insert(value)
	}
	
	// Test searching for existing values
	for _, value := range values {
		if !bst.Search(value) {
			t.Errorf("Expected to find value %d, but search returned false", value)
		}
	}
	
	// Test searching for non-existing values
	nonExisting := []int{10, 25, 35, 45, 55, 65, 75, 85, 100}
	for _, value := range nonExisting {
		if bst.Search(value) {
			t.Errorf("Expected not to find value %d, but search returned true", value)
		}
	}
	
	// Test searching in empty tree
	emptyBST := NewBinarySearchTree()
	if emptyBST.Search(50) {
		t.Errorf("Expected search in empty tree to return false")
	}
}

func TestBinarySearchTree_Delete(t *testing.T) {
	bst := NewBinarySearchTree()
	values := []int{50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45}
	
	// Insert values
	for _, value := range values {
		bst.Insert(value)
	}
	
	originalSize := bst.Size()
	
	// Test deleting leaf node
	bst.Delete(10)
	if bst.Search(10) {
		t.Errorf("Expected value 10 to be deleted")
	}
	if bst.Size() != originalSize-1 {
		t.Errorf("Expected size to be %d after deletion, got %d", originalSize-1, bst.Size())
	}
	
	// Test deleting node with one child
	bst.Delete(20)
	if bst.Search(20) {
		t.Errorf("Expected value 20 to be deleted")
	}
	
	// Test deleting node with two children
	bst.Delete(30)
	if bst.Search(30) {
		t.Errorf("Expected value 30 to be deleted")
	}
	
	// Test deleting non-existing value
	originalSize = bst.Size()
	bst.Delete(999)
	if bst.Size() != originalSize {
		t.Errorf("Expected size to remain %d after deleting non-existing value, got %d", originalSize, bst.Size())
	}
	
	// Test deleting from empty tree
	emptyBST := NewBinarySearchTree()
	emptyBST.Delete(50) // Should not panic
}

func TestBinarySearchTree_Traversals(t *testing.T) {
	bst := NewBinarySearchTree()
	values := []int{50, 30, 70, 20, 40, 60, 80}
	
	// Insert values
	for _, value := range values {
		bst.Insert(value)
	}
	
	// Test inorder traversal (should be sorted)
	expectedInorder := []int{20, 30, 40, 50, 60, 70, 80}
	inorder := bst.InorderTraversal()
	if !reflect.DeepEqual(inorder, expectedInorder) {
		t.Errorf("Expected inorder %v, got %v", expectedInorder, inorder)
	}
	
	// Test preorder traversal
	expectedPreorder := []int{50, 30, 20, 40, 70, 60, 80}
	preorder := bst.PreorderTraversal()
	if !reflect.DeepEqual(preorder, expectedPreorder) {
		t.Errorf("Expected preorder %v, got %v", expectedPreorder, preorder)
	}
	
	// Test postorder traversal
	expectedPostorder := []int{20, 40, 30, 60, 80, 70, 50}
	postorder := bst.PostorderTraversal()
	if !reflect.DeepEqual(postorder, expectedPostorder) {
		t.Errorf("Expected postorder %v, got %v", expectedPostorder, postorder)
	}
	
	// Test level-order traversal
	expectedLevelOrder := []int{50, 30, 70, 20, 40, 60, 80}
	levelOrder := bst.LevelOrderTraversal()
	if !reflect.DeepEqual(levelOrder, expectedLevelOrder) {
		t.Errorf("Expected level-order %v, got %v", expectedLevelOrder, levelOrder)
	}
	
	// Test traversals on empty tree
	emptyBST := NewBinarySearchTree()
	if len(emptyBST.InorderTraversal()) != 0 {
		t.Errorf("Expected empty inorder traversal")
	}
	if len(emptyBST.LevelOrderTraversal()) != 0 {
		t.Errorf("Expected empty level-order traversal")
	}
}

func TestBinarySearchTree_Height(t *testing.T) {
	bst := NewBinarySearchTree()
	
	// Test height of empty tree
	if bst.Height() != -1 {
		t.Errorf("Expected height of empty tree to be -1, got %d", bst.Height())
	}
	
	// Test height of single node
	bst.Insert(50)
	if bst.Height() != 0 {
		t.Errorf("Expected height of single node to be 0, got %d", bst.Height())
	}
	
	// Test height of balanced tree
	values := []int{50, 30, 70, 20, 40, 60, 80}
	for _, value := range values {
		bst.Insert(value)
	}
	expectedHeight := 2
	if bst.Height() != expectedHeight {
		t.Errorf("Expected height %d, got %d", expectedHeight, bst.Height())
	}
	
	// Test height of unbalanced tree
	unbalancedBST := NewBinarySearchTree()
	for i := 1; i <= 5; i++ {
		unbalancedBST.Insert(i)
	}
	expectedUnbalancedHeight := 4
	if unbalancedBST.Height() != expectedUnbalancedHeight {
		t.Errorf("Expected unbalanced height %d, got %d", expectedUnbalancedHeight, unbalancedBST.Height())
	}
}

func TestBinarySearchTree_Size(t *testing.T) {
	bst := NewBinarySearchTree()
	
	// Test size of empty tree
	if bst.Size() != 0 {
		t.Errorf("Expected size of empty tree to be 0, got %d", bst.Size())
	}
	
	// Test size after insertions
	values := []int{50, 30, 70, 20, 40, 60, 80}
	for i, value := range values {
		bst.Insert(value)
		expectedSize := i + 1
		if bst.Size() != expectedSize {
			t.Errorf("Expected size %d after inserting %d, got %d", expectedSize, value, bst.Size())
		}
	}
	
	// Test size after deletions
	bst.Delete(20)
	expectedSize := 6
	if bst.Size() != expectedSize {
		t.Errorf("Expected size %d after deletion, got %d", expectedSize, bst.Size())
	}
}

func TestBinarySearchTree_IsBalanced(t *testing.T) {
	bst := NewBinarySearchTree()
	
	// Test empty tree is balanced
	if !bst.IsBalanced() {
		t.Errorf("Expected empty tree to be balanced")
	}
	
	// Test single node is balanced
	bst.Insert(50)
	if !bst.IsBalanced() {
		t.Errorf("Expected single node to be balanced")
	}
	
	// Test balanced tree
	values := []int{50, 30, 70, 20, 40, 60, 80}
	for _, value := range values {
		bst.Insert(value)
	}
	if !bst.IsBalanced() {
		t.Errorf("Expected balanced tree to be balanced")
	}
	
	// Test unbalanced tree
	unbalancedBST := NewBinarySearchTree()
	for i := 1; i <= 5; i++ {
		unbalancedBST.Insert(i)
	}
	if unbalancedBST.IsBalanced() {
		t.Errorf("Expected unbalanced tree to not be balanced")
	}
}

func TestBinarySearchTree_IsEmpty(t *testing.T) {
	bst := NewBinarySearchTree()
	
	// Test empty tree
	if !bst.IsEmpty() {
		t.Errorf("Expected new tree to be empty")
	}
	
	// Test non-empty tree
	bst.Insert(50)
	if bst.IsEmpty() {
		t.Errorf("Expected tree with nodes to not be empty")
	}
}

func TestBinarySearchTree_Clear(t *testing.T) {
	bst := NewBinarySearchTree()
	values := []int{50, 30, 70, 20, 40, 60, 80}
	
	// Insert values
	for _, value := range values {
		bst.Insert(value)
	}
	
	// Verify tree is not empty
	if bst.IsEmpty() {
		t.Errorf("Expected tree to not be empty before clear")
	}
	
	// Clear tree
	bst.Clear()
	
	// Verify tree is empty
	if !bst.IsEmpty() {
		t.Errorf("Expected tree to be empty after clear")
	}
	if bst.Size() != 0 {
		t.Errorf("Expected size to be 0 after clear, got %d", bst.Size())
	}
	if bst.Root != nil {
		t.Errorf("Expected root to be nil after clear")
	}
}

func TestBinarySearchTree_String(t *testing.T) {
	bst := NewBinarySearchTree()
	
	// Test empty tree string representation
	emptyStr := bst.String()
	if emptyStr != "Empty BST" {
		t.Errorf("Expected 'Empty BST', got '%s'", emptyStr)
	}
	
	// Test non-empty tree string representation
	bst.Insert(50)
	bst.Insert(30)
	bst.Insert(70)
	
	str := bst.String()
	if str == "" {
		t.Errorf("Expected non-empty string representation")
	}
}

// Benchmark tests
func BenchmarkBinarySearchTree_Insert(b *testing.B) {
	bst := NewBinarySearchTree()
	for i := 0; i < b.N; i++ {
		bst.Insert(i)
	}
}

func BenchmarkBinarySearchTree_Search(b *testing.B) {
	bst := NewBinarySearchTree()
	// Insert 1000 values
	for i := 0; i < 1000; i++ {
		bst.Insert(i)
	}
	
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		bst.Search(i % 1000)
	}
}

func BenchmarkBinarySearchTree_Delete(b *testing.B) {
	bst := NewBinarySearchTree()
	// Insert 1000 values
	for i := 0; i < 1000; i++ {
		bst.Insert(i)
	}
	
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		bst.Delete(i % 1000)
	}
}

File: algorithms/code/tests/BinarySearchTreeTest.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Arrays;
import java.util.List;

/**
 * Comprehensive test suite for BinarySearchTree implementation
 */
public class BinarySearchTreeTest {
    
    private BinarySearchTree bst;
    
    @BeforeEach
    void setUp() {
        bst = new BinarySearchTree();
    }
    
    @Test
    @DisplayName("Test inserting into empty tree")
    void testInsertIntoEmptyTree() {
        bst.insert(50);
        assertNotNull(bst.root, "Root should not be null after insertion");
        assertEquals(50, bst.root.value, "Root value should be 50");
    }
    
    @Test
    @DisplayName("Test inserting smaller and larger values")
    void testInsertSmallerAndLargerValues() {
        bst.insert(50);
        bst.insert(30);
        bst.insert(70);
        
        assertNotNull(bst.root.left, "Left child should not be null");
        assertEquals(30, bst.root.left.value, "Left child value should be 30");
        
        assertNotNull(bst.root.right, "Right child should not be null");
        assertEquals(70, bst.root.right.value, "Right child value should be 70");
    }
    
    @Test
    @DisplayName("Test inserting duplicate values")
    void testInsertDuplicateValues() {
        bst.insert(50);
        int originalSize = bst.size();
        
        bst.insert(50); // Insert duplicate
        
        assertEquals(originalSize, bst.size(), "Size should not change after inserting duplicate");
    }
    
    @Test
    @DisplayName("Test searching for existing values")
    void testSearchExistingValues() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        // Test searching for existing values
        for (int value : values) {
            assertTrue(bst.search(value), "Should find value " + value);
        }
    }
    
    @Test
    @DisplayName("Test searching for non-existing values")
    void testSearchNonExistingValues() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        int[] nonExisting = {10, 25, 35, 45, 55, 65, 75, 85, 100};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        // Test searching for non-existing values
        for (int value : nonExisting) {
            assertFalse(bst.search(value), "Should not find value " + value);
        }
    }
    
    @Test
    @DisplayName("Test searching in empty tree")
    void testSearchInEmptyTree() {
        assertFalse(bst.search(50), "Search in empty tree should return false");
    }
    
    @Test
    @DisplayName("Test deleting leaf node")
    void testDeleteLeafNode() {
        int[] values = {50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        int originalSize = bst.size();
        
        // Delete leaf node
        bst.delete(10);
        
        assertFalse(bst.search(10), "Value 10 should be deleted");
        assertEquals(originalSize - 1, bst.size(), "Size should decrease by 1");
    }
    
    @Test
    @DisplayName("Test deleting node with one child")
    void testDeleteNodeWithOneChild() {
        int[] values = {50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        // Delete node with one child
        bst.delete(20);
        
        assertFalse(bst.search(20), "Value 20 should be deleted");
    }
    
    @Test
    @DisplayName("Test deleting node with two children")
    void testDeleteNodeWithTwoChildren() {
        int[] values = {50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        // Delete node with two children
        bst.delete(30);
        
        assertFalse(bst.search(30), "Value 30 should be deleted");
    }
    
    @Test
    @DisplayName("Test deleting non-existing value")
    void testDeleteNonExistingValue() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        int originalSize = bst.size();
        
        // Delete non-existing value
        bst.delete(999);
        
        assertEquals(originalSize, bst.size(), "Size should not change after deleting non-existing value");
    }
    
    @Test
    @DisplayName("Test deleting from empty tree")
    void testDeleteFromEmptyTree() {
        // Should not throw exception
        assertDoesNotThrow(() -> bst.delete(50), "Deleting from empty tree should not throw exception");
    }
    
    @Test
    @DisplayName("Test inorder traversal")
    void testInorderTraversal() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        List<Integer> expected = Arrays.asList(20, 30, 40, 50, 60, 70, 80);
        List<Integer> actual = bst.inorderTraversal();
        
        assertEquals(expected, actual, "Inorder traversal should be sorted");
    }
    
    @Test
    @DisplayName("Test preorder traversal")
    void testPreorderTraversal() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        List<Integer> expected = Arrays.asList(50, 30, 20, 40, 70, 60, 80);
        List<Integer> actual = bst.preorderTraversal();
        
        assertEquals(expected, actual, "Preorder traversal should match expected order");
    }
    
    @Test
    @DisplayName("Test postorder traversal")
    void testPostorderTraversal() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        List<Integer> expected = Arrays.asList(20, 40, 30, 60, 80, 70, 50);
        List<Integer> actual = bst.postorderTraversal();
        
        assertEquals(expected, actual, "Postorder traversal should match expected order");
    }
    
    @Test
    @DisplayName("Test level-order traversal")
    void testLevelOrderTraversal() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        List<Integer> expected = Arrays.asList(50, 30, 70, 20, 40, 60, 80);
        List<Integer> actual = bst.levelOrderTraversal();
        
        assertEquals(expected, actual, "Level-order traversal should match expected order");
    }
    
    @Test
    @DisplayName("Test traversals on empty tree")
    void testTraversalsOnEmptyTree() {
        assertTrue(bst.inorderTraversal().isEmpty(), "Inorder traversal of empty tree should be empty");
        assertTrue(bst.preorderTraversal().isEmpty(), "Preorder traversal of empty tree should be empty");
        assertTrue(bst.postorderTraversal().isEmpty(), "Postorder traversal of empty tree should be empty");
        assertTrue(bst.levelOrderTraversal().isEmpty(), "Level-order traversal of empty tree should be empty");
    }
    
    @Test
    @DisplayName("Test height of empty tree")
    void testHeightOfEmptyTree() {
        assertEquals(-1, bst.height(), "Height of empty tree should be -1");
    }
    
    @Test
    @DisplayName("Test height of single node")
    void testHeightOfSingleNode() {
        bst.insert(50);
        assertEquals(0, bst.height(), "Height of single node should be 0");
    }
    
    @Test
    @DisplayName("Test height of balanced tree")
    void testHeightOfBalancedTree() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        
        for (int value : values) {
            bst.insert(value);
        }
        
        assertEquals(2, bst.height(), "Height of balanced tree should be 2");
    }
    
    @Test
    @DisplayName("Test height of unbalanced tree")
    void testHeightOfUnbalancedTree() {
        // Create unbalanced tree (all nodes to the right)
        for (int i = 1; i <= 5; i++) {
            bst.insert(i);
        }
        
        assertEquals(4, bst.height(), "Height of unbalanced tree should be 4");
    }
    
    @Test
    @DisplayName("Test size operations")
    void testSizeOperations() {
        // Test size of empty tree
        assertEquals(0, bst.size(), "Size of empty tree should be 0");
        
        // Test size after insertions
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        for (int i = 0; i < values.length; i++) {
            bst.insert(values[i]);
            assertEquals(i + 1, bst.size(), "Size should be " + (i + 1) + " after inserting " + values[i]);
        }
        
        // Test size after deletion
        bst.delete(20);
        assertEquals(6, bst.size(), "Size should be 6 after deletion");
    }
    
    @Test
    @DisplayName("Test balance checking")
    void testBalanceChecking() {
        // Test empty tree is balanced
        assertTrue(bst.isBalanced(), "Empty tree should be balanced");
        
        // Test single node is balanced
        bst.insert(50);
        assertTrue(bst.isBalanced(), "Single node should be balanced");
        
        // Test balanced tree
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        for (int value : values) {
            bst.insert(value);
        }
        assertTrue(bst.isBalanced(), "Balanced tree should be balanced");
        
        // Test unbalanced tree
        BinarySearchTree unbalancedBST = new BinarySearchTree();
        for (int i = 1; i <= 5; i++) {
            unbalancedBST.insert(i);
        }
        assertFalse(unbalancedBST.isBalanced(), "Unbalanced tree should not be balanced");
    }
    
    @Test
    @DisplayName("Test empty checking")
    void testEmptyChecking() {
        // Test empty tree
        assertTrue(bst.isEmpty(), "New tree should be empty");
        
        // Test non-empty tree
        bst.insert(50);
        assertFalse(bst.isEmpty(), "Tree with nodes should not be empty");
    }
    
    @Test
    @DisplayName("Test clear operation")
    void testClearOperation() {
        int[] values = {50, 30, 70, 20, 40, 60, 80};
        
        // Insert values
        for (int value : values) {
            bst.insert(value);
        }
        
        // Verify tree is not empty
        assertFalse(bst.isEmpty(), "Tree should not be empty before clear");
        
        // Clear tree
        bst.clear();
        
        // Verify tree is empty
        assertTrue(bst.isEmpty(), "Tree should be empty after clear");
        assertEquals(0, bst.size(), "Size should be 0 after clear");
        assertNull(bst.root, "Root should be null after clear");
    }
    
    @Test
    @DisplayName("Test string representation")
    void testStringRepresentation() {
        // Test empty tree string representation
        String emptyStr = bst.toString();
        assertEquals("Empty BST", emptyStr, "Empty tree should return 'Empty BST'");
        
        // Test non-empty tree string representation
        bst.insert(50);
        bst.insert(30);
        bst.insert(70);
        
        String str = bst.toString();
        assertNotNull(str, "String representation should not be null");
        assertFalse(str.isEmpty(), "String representation should not be empty");
    }
    
    @Test
    @DisplayName("Test complex operations sequence")
    void testComplexOperationsSequence() {
        // Insert multiple values
        int[] values = {50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45};
        for (int value : values) {
            bst.insert(value);
        }
        
        // Verify initial state
        assertEquals(11, bst.size(), "Initial size should be 11");
        assertTrue(bst.isBalanced(), "Tree should be balanced");
        
        // Perform various operations
        assertTrue(bst.search(40), "Should find value 40");
        assertFalse(bst.search(90), "Should not find value 90");
        
        // Delete some values
        bst.delete(10);
        bst.delete(30);
        bst.delete(70);
        
        // Verify final state
        assertEquals(8, bst.size(), "Final size should be 8");
        assertFalse(bst.search(10), "Should not find deleted value 10");
        assertFalse(bst.search(30), "Should not find deleted value 30");
        assertFalse(bst.search(70), "Should not find deleted value 70");
        
        // Verify traversals still work
        List<Integer> inorder = bst.inorderTraversal();
        assertFalse(inorder.isEmpty(), "Inorder traversal should not be empty");
        assertTrue(inorder.size() == 8, "Inorder traversal should have 8 elements");
    }
}
